cost/azure/rightsize_compute_instances/azure_compute_rightsizing.pt

```diff
@@ -706,29 +737,82 @@ script "js_azure_instances_metrics", type: "javascript" do
   start_date.setSeconds(0)
   start_date.setMinutes(0)
 
-  timespan = start_date.toISOString() + "/" + end_date.toISOString()
+  function get_host_by_region(region) {
+    if (param_azure_endpoint === "management.chinacloudapi.cn") {
+      // China region has different endpoint for Metric getBatch API
+      return region + ".metrics.monitor.azure.cn";
+    }
+    else
+      return region + ".metrics.monitor.azure.com";
+  }
+
+  var resourceIDs = _.map(ds_azure_instances_batch, function(instance) {
+    return instance.resourceID;
+  });
 
   var request = {
     auth: "auth_azure",
     pagination: "pagination_azure",
-    host: param_azure_endpoint,
-    path: resourceID + "/providers/microsoft.insights/metrics",
+    host: get_host_by_region(ds_azure_instances_batch[0].region),
+    verb: "POST",
+    path: "subscriptions/" + ds_azure_instances_batch[0].subscriptionId + "/metrics:getBatch",
     query_params: {
-      "api-version": "2018-01-01",
-      "timespan": timespan,
+      "api-version": "2023-10-01",
+      "starttime": start_date.toISOString(),
+      "endtime": end_date.toISOString(),
       "interval": ds_stats_interval["api"],
       "metricnames": "Percentage CPU,Available Memory Percentage",
+      "metricnamespace": "Microsoft.Compute/VirtualMachines",
       "aggregation": "average,maximum,minimum"
     },
     headers: {
-      "User-Agent": "RS Policies"
+      "User-Agent": "RS Policies",
+      "Content-Type": "application/json"
     },
-    // Ignore status 400, 403, and 404 which can be returned in certain (legacy) types of Azure Subscriptions
-    ignore_status: [400, 403, 404]
+    body: JSON.stringify({"resourceids": resourceIDs})
   }
 EOS
 end
 
+datasource "ds_azure_instances_metrics" do
+  batch true
+  run_script $js_azure_instances_metrics, $ds_azure_metrics_batch, $ds_azure_instances_status_filtered
+end
+
+script "js_azure_instances_metrics", type: "javascript" do
+  parameters "ds_azure_metrics_batch", "ds_azure_instances_status_filtered"
+  result "result"
+  code <<-EOS
+
+  var instances_map = {};
+  _.each(ds_azure_instances_status_filtered, function(instance) {
+    instances_map[instance.resourceID] = instance
+  })
```

Comment:
Can use _.groupBy() function in future for this - https://underscorejs.org/#groupBy